lib/date_range.rb

```diff
@@ -0,0 +1,27 @@
+module BookingSystem
+  class DateRange
+
```

Comment:
All the places that you create a `DateRange`, it seems like you're only using it as a tool to get an array of dates. This indicates that a method might be a better choice here, one that takes a checkin and checkout date and returns an array of dates.

The other option would be to write an instance method, something like `DateRange#overlap?(other)`, that takes another `DateRange` and checks to see if they overlap. Then this complex bit of logic would be hidden away in `DateRange` instead of sitting in `Hotel` or `Reservation`.

As long as you've identified "manage a range of dates" as a responsibility, then it probably makes sense to do the later, and keep that responsibility out of the `Hotel` class.

This would also make it much easier to test!